### DISCUSSION TITLE
I have a question about Equal's implementation.

### DISCUSSION THREAD

**Discussion by umaxyon** [type-challenges/type-challenges#5346] [Q&A]
Hi. I am a beginner.

The implementation of "Equal" in type-challenges is as follows, 
```typescript
export type Equal<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false
```
but I don't know why this works.
What goes into the "T" and when is the function executed?

Why not the following?
```typescript
export type Equal<X, Y> = X extends Y ? true : false
```
Please tell me. thank you.

**Comment by yangliguo7** [comment]
This can be simply understood as the difference between `==` and `===`  (PS:I just want to explain it simply)
consider such a problem
```
type A  = { x : 1 , y : 1}  extends { x : 1 } ? 1 : 2  
A ==> 1 
but { x : 1 , y : 1}  not equal  { x : 1 } 
```

**Comment by zjlovezj** [comment]
I did a little search.
There is no perfect solution for Equal<X, y> implementation.

see
https://github.com/Microsoft/TypeScript/issues/27024#issuecomment-421529650

**Comment by sbr61** [comment]
See my comment an [Could you explain how does Equal type implementation work?](https://github.com/type-challenges/type-challenges/discussions/9100) for a detailed explanation.
